src/dialer.js

```diff
@@ -29,105 +29,88 @@ class Dialer {
     transportManager,
     peerStore,
     concurrency = MAX_PARALLEL_DIALS,
-    timeout = DIAL_TIMEOUT
+    timeout = DIAL_TIMEOUT,
+    perPeerLimit = MAX_PER_PEER_DIALS
   }) {
     this.transportManager = transportManager
     this.peerStore = peerStore
     this.concurrency = concurrency
     this.timeout = timeout
-    this.queue = new PQueue({ concurrency, timeout, throwOnTimeout: true })
-
-    /**
-     * @property {IdentifyService}
-     */
-    this._identifyService = null
-  }
-
-  set identifyService (service) {
-    this._identifyService = service
+    this.perPeerLimit = perPeerLimit
+    this.tokens = [...new Array(concurrency)].map((_, index) => index)
   }
 
   /**
-   * @type {IdentifyService}
-   */
-  get identifyService () {
-    return this._identifyService
-  }
-
-  /**
-   * Connects to a given `Multiaddr`. `addr` should include the id of the peer being
-   * dialed, it will be used for encryption verification.
+   * Connects to the first success of a given list of `Multiaddr`. `addrs` should
+   * include the id of the peer being dialed, it will be used for encryption verification.
    *
-   * @async
-   * @param {Multiaddr} addr The address to dial
+   * @param {Array<Multiaddr>|Multiaddr} addrs
    * @param {object} [options]
    * @param {AbortSignal} [options.signal] An AbortController signal
    * @returns {Promise<Connection>}
    */
-  async connectToMultiaddr (addr, options = {}) {
-    addr = multiaddr(addr)
-    let conn
-    let controller
+  async connectToMultiaddr (addrs, options = {}) {
+    if (!Array.isArray(addrs)) addrs = [multiaddr(addrs)]
 
-    if (!options.signal) {
-      controller = new AbortController()
-      options.signal = controller.signal
+    const dialAction = (addr, options) => {
+      if (options.signal.aborted) throw errCode(new Error('already aborted'), 'ERR_ALREADY_ABORTED')
+      return this.transportManager.dial(addr, options)
     }
+    const dialRequest = new DialRequest({
+      addrs,
+      dialAction,
+      dialer: this
+    })
+
+    // Combine the timeout signal and options.signal, if provided
+    const timeoutController = new AbortController()
+    const signals = [timeoutController.signal]
+    options.signal && signals.push(options.signal)
+    const signal = anySignal(signals)
+    const timeoutId = setTimeout(() => timeoutController.abort(), this.timeout)
 
     try {
-      conn = await this.queue.add(() => this.transportManager.dial(addr, options))
+      const dialResult = await dialRequest.run({ ...options, signal })
+      clearTimeout(timeoutId)
+      return dialResult
```

Comment:
Should we log success dial?
Also, maybe do the return after the catch instead, so that it is easier to find what is returned